modules/widgets/src/pure-js/widget-impl.ts

```diff
@@ -0,0 +1,95 @@
+// deck.gl
+// SPDX-License-Identifier: MIT
+// Copyright (c) vis.gl contributors
+
+/* global document */
+import {_deepEqual, _applyStyles, _removeStyles} from '@deck.gl/core';
+import type {Deck, Widget, WidgetPlacement} from '@deck.gl/core';
+
+export type WidgetImplProps = {
+  id?: string;
+  /** Widget positioning within the view. Default 'top-left'. */
+  placement?: WidgetPlacement;
+  /** CSS inline style overrides. */
+  style?: Partial<CSSStyleDeclaration>;
+  /** Additional CSS class. */
+  className?: string;
+};
+export abstract class WidgetImpl<PropsT extends WidgetImplProps> implements Widget<PropsT> {
+  id: string;
+  props: Required<PropsT>;
+
+  deck?: Deck<any>;
+  element?: HTMLDivElement;
+
+  static defaultProps: Required<WidgetImplProps> = {
+    id: 'widget',
+    placement: 'top-left',
+    style: {},
+    className: ''
+  };
+
+  abstract className: Required<string>;
+
+  constructor(props: Required<PropsT>) {
+    this.id = props.id || 'widget';
+    this.props = props;
+  }
+
+  abstract onRenderHTML(): void;
+
+  onAdd({deck}: {deck: Deck<any>}): HTMLDivElement {
+    this.deck = deck;
+    const {style, className} = this.props;
+    const el = this._createRootElement({
+      widgetClassName: this.className,
+      className,
+      style
+    });
+    this.element = el;
+    this.onRenderHTML();
+    return this.element;
+  }
+
+  onRemove() {
+    this.deck = undefined;
+    this.element = undefined;
+  }
+
+  setProps(props: Partial<PropsT>) {
+    const oldProps = this.props;
+    const el = this.element;
+    if (el) {
+      if (oldProps.className !== props.className) {
+        if (oldProps.className) el.classList.remove(oldProps.className);
+        if (props.className) el.classList.add(props.className);
+      }
+
+      if (!_deepEqual(oldProps.style, props.style, 1)) {
+        _removeStyles(el, oldProps.style);
+        _applyStyles(el, props.style);
+      }
+    }
+
+    Object.assign(this.props, props);
+    this.onRenderHTML();
+  }
+
+  _createRootElement(props: {
+    widgetClassName: string;
+    className?: string;
+    style?: Partial<CSSStyleDeclaration>;
+  }) {
+    const {widgetClassName, className, style} = props;
+    const element = document.createElement('div');
+    ['deck-widget', widgetClassName, className]
+      .filter(Boolean)
+      .forEach(className => element.classList.add(className));
+    _applyStyles(element, style);
+    return element;
+  }
+
+  _applyStyles(el: HTMLElement, style: Partial<CSSStyleDeclaration>) {
+    _applyStyles(el, style);
+  }
```

Comment:
Is this necessary?

Reply:
If it use useful to users then a little nicer than asking them to import internal functions. But perhaps we already handle it in setProps now, so we don't need to expose it.